pkg/mutation/schema/node.go

```diff
@@ -0,0 +1,241 @@
+		// No child of the key and type exists.
+		return fmt.Errorf("no type %q for key %q: %w", childType, childKey, ErrNotFound)
+	}
+
+	child := n.Children[childKey][childType]
+	err := child.Remove(id, path[1:])
+	if err != nil {
+		return fmt.Errorf("%q: %w", childKey, err)
+	}
+
+	// Delete the type from the child if it is no longer referenced.
+	if len(child.ReferencedBy) == 0 {
+		// No references to this child of this type exist.
+		delete(n.Children[childKey], childType)
+	} else {
+		n.Children[childKey][childType] = child
+	}
+
+	// Delete the child if it is no longer referenced.
+	if len(n.Children[childKey]) == 0 {
+		delete(n.Children, childKey)
+	}
+
+	return nil
+}
+
+func (n *node) conflicts(childKey string) idSet {
+	conflicts := make(idSet)
+
+	// Count the number of distinct types with this key.
+	nTypes := len(n.Children[childKey])
+	if _, hasUnknown := n.Children[childKey][unknown]; hasUnknown {
+		// Nodes whose types we are unable to determine do not count against this
+		// check.
+		nTypes--
+	}
+
+	for nodeType, child := range n.Children[childKey] {
+		if nodeType == unknown {
+			// If we don't know the type of a node, we assume it conflicts with nothing.
+			continue
+		}
+		// There are conflicts if either:
+		// 1) there are more than one non-unknown types for the Child, or
+		// 2) the Child is a List and defines multiple keys.
+		if nTypes > 1 || nodeType == parser.ListNode && len(child.Children) > 1 {
+			conflicts = merge(conflicts, child.ReferencedBy)
+		}
+	}
+
+	// If more than 1 non-unknown types are declared, this node is part of a
+	// schema conflict.
+	return conflicts
+}
+
+// HasConflicts returns true if there are any schema type conflicts along the
+// passed path.
+//
+// Returns an error if the path does not exist.
+func (n *node) HasConflicts(path []parser.Node) (bool, error) {
+	if len(path) == 0 {
+		return false, nil
+	}
+
+	childKey := path[0].Key()
+	childType := headType(path)
+	if _, found := n.Children[childKey]; !found {
+		// Path has not been added, so there can be no conflicts.
+		return false, ErrNotFound
+	}
+
+	// Count the number of distinct types with this key.
+	if len(n.conflicts(childKey)) > 0 {
+		return true, nil
+	}
+
+	if _, found := n.Children[childKey][childType]; !found {
+		// Path has not been added, so there can be no conflicts.
+		return false, ErrNotFound
+	}
+	child := n.Children[childKey][childType]
+	return child.HasConflicts(path[1:])
+}
+
+// merge inserts elements from `from` into `into`. Returns `into`, or a
+// reference to a new map if `into` is nil.
+func merge(into, from idSet) idSet {
+	if len(into) == 0 && len(from) == 0 {
+		return nil
+	}
+	if into == nil {
+		into = make(idSet)
+	}
+	for k := range from {
+		into[k] = true
+	}
+	return into
+}
+
+// headType returns the type of the second Node, if it exists.
+// This is essential for determining whether the current location in a schema
+// path is a list.
```

Comment:
No, it can't and it does not fail. Many of the unit tests deal with exactly this case and prove it works as intended. The problem is that path.Path stores a list as two nodes [Object Node, List Node] and we care that the first node's children all have the same type.

Reply:
NVM, I think I misread how this was used.